### PR TITLE
fix(color-contrast): Prevent crash on large inline elments #1306

### LINKED ISSUE #1259: Runtime issue: TypeError: Cannot read property '0' of undefined
<< Please describe how you encountered this issue >>

```
TypeError: Cannot read property '0' of undefined
    at chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:109:137989
    at Array.every (<anonymous>)
    at chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:109:137957
    at Array.forEach (<anonymous>)
    at Object.a.filteredRectStack (chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:109:137901)
    at Object.a.getBackgroundStack (chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:109:138215)
    at Object.a.getBackgroundColor (chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:109:139420)
    at Object.evaluate (chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:109:68263)
    at b.run (chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:76:8610)
    at chrome-extension://lhdoppojpmngadmnindnejefpokejbdd/adapter.bundle.js:76:12663
```

### LINKED ISSUE #1306: Runtime issue: TypeError: Cannot read property '0' of undefined
I got the following error running axe on https://www.luckyscript.me/

```
TypeError: Cannot read property '0' of undefined
    at chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:68:165689
    at Array.every (<anonymous>)
    at chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:68:165657
    at Array.forEach (<anonymous>)
    at Object.i.filteredRectStack (chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:68:165601)
    at Object.i.getBackgroundStack (chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:68:165915)
    at Object.i.getBackgroundColor (chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:68:167120)
    at Object.evaluate (chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:68:96067)
    at v.run (chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:37:8511)
    at chrome-extension://lfmcehohgifnaodaogknapedjiaoebgo/adapter.bundle.js:37:12495
```

### PR DESCRIPTION
**Pull request by WilcoFiers**
Worked out the problem behind this issue that Stephen created a patch for 2 weeks ago. Basically, inline elements can have multiple rects when they wrap over multiple lines. This can happen in 2 cases. Either the inline element has rects that don't fit the viewport at all (which is very unlikely), or axe-core doesn't scroll the entire element into view (which was the case for the reported issue).

The reason why axe-core doesn't always scroll the entire element into view is because we don't use `.scrollIntoView(true)` (alignToTop) most of the time. This is to avoid scrolling elements with overflow: hidden. There is probably a way to avoid this, and scroll the entire element into view, but that would require a lot more work. Probably not worth the effort, given how uncommon this issue is.

Closes #1306, closes #1259

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [x] Follows the commit message policy, appropriate for next version
- [x] Has documentation updated, a DU ticket, or requires no documentation change
- [x] Includes new tests, or was unnecessary
- [x] Code is reviewed for security by: dylanb

### CODE REVIEW

File: lib/commons/color/get-background-color.js [hunk 1]
```diff
@@ -231,33 +231,42 @@ color.getCoords = function(rect) {
...
+			return rect.width && rect.width > 0;
+		})
+		.map(rect => {
+			let coords = color.getCoords(rect);
+			if (coords) {
+				return dom.shadowElementsFromPoint(coords.x, coords.y);
+			}
+		});
+
+	if (filteredArr.some(stack => stack === undefined)) {
```
Comment (WilcoFiers) [inline lib/commons/color/get-background-color.js]: Mostly a return-early refactor, these 3 lines are the new bit.

File: lib/commons/color/get-background-color.js [hunk 2]
```diff
@@ -351,7 +352,8 @@ color.getBackgroundColor = function(elm, bgElms = [], noScroll = false) {
 	if (noScroll !== true) {
 		// Avoid scrolling overflow:hidden containers, by only aligning to top
 		// when not doing so would move the center point above the viewport top.
-		const alignToTop = elm.clientHeight - 2 >= window.innerHeight * 2;
+		const clientHeight = elm.getBoundingClientRect().height;
+		const alignToTop = clientHeight - 2 >= window.innerHeight * 2;
```
Comment (WilcoFiers) [inline lib/commons/color/get-background-color.js]: `clientHeight` is 0 on inline elements with multiple rects. getBoundingClientRect does give the rect for the entire thing.

File: lib/commons/color/get-background-color.js [hunk 3]
```diff
@@ -231,33 +231,42 @@ color.getCoords = function(rect) {
...
-						return dom.shadowElementsFromPoint(coords.x, coords.y);
-					}
-				});
-			// add bounding client rect stack for comparison later
-			filteredArr.splice(0, 0, boundingStack);
-			return filteredArr;
-		} else {
-			return [boundingStack];
-		}
+	if (!boundingCoords) {
```
Comment (stephenmathieson) [inline lib/commons/color/get-background-color.js]: 👍 thank you for this. Indenting the happy path always bugs me.

File: lib/commons/color/get-background-color.js [hunk 4]
```diff
@@ -231,33 +231,42 @@ color.getCoords = function(rect) {
...
+	// If the element does not have multiple rects, like for display:block, return a single stack
+	if (!rects || rects.length <= 1) {
+		return [boundingStack];
+	}
+
+	// Handle inline elements spanning multiple lines to be evaluated
+	let filteredArr = rects
+		.filter(rect => {
+			// exclude manual line breaks in Chrome/Safari
+			return rect.width && rect.width > 0;
```
Comment (dylanb) [inline lib/commons/color/get-background-color.js]: nit: Why not just `return rect.width > 0`?

File: lib/commons/color/get-background-color.js (hunk 1 above)
Comment (dylanb) [reply lib/commons/color/get-background-color.js]: Why not filter these out?

File: lib/commons/color/get-background-color.js (hunk 4 above)
Comment (WilcoFiers) [reply lib/commons/color/get-background-color.js]: We could. I just changed indentation on this. Didn't actually look to refactor.

**Pull request merged**